Support auto generate redistribute cost

zpcore opened the pull request:
This PR supports automatically generate the `redistribute_cost` for OpSpec if developer didn't specify the attribute when writing the sharding propagation strategy. This can simply the code for `register_op_strategy` for operators since we don't need to write the computation logic for `redistribute_cost` anymore.

The change is made in the decorator `ShardingPropagator.register_op_strategy`, which should not impact @fmassa 's auto parallel. The code also assigns the tensor_meta data if the filed is missing. 

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

zpcore committed:
Support auto generate redistribute cost

zpcore committed:
clear global var

wanchaol reviewed (commented):
first pass, please see inlined comments

zpcore commented on test/distributed/tensor/test_strategy_registration.py:
```diff
@@ -0,0 +1,201 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
```
Done!

zpcore committed:
refactor out func

zpcore committed:
remove file

zpcore commented:
Per offline discussion with @XilunWu , the PR is not a high priority since we can quickly add those missing strategies to unblock others. Let me check how many costs are missing and we may drop the PR based on the finding.

zpcore commented:
Did a quick coverage run to detect how many ops that miss the OpSpec info:
- redistribute cost missing:
```
aten::detach
aten::bernoulli
aten::_to_copy
aten::bucketize.Tensor
aten::stack
aten::clone
aten::cat
aten::split.Tensor
aten::t
aten::equal
aten::slice.Tensor
aten::zeros_like
aten::copy_
aten::empty_like
aten::full_like
aten::select.int
aten::bernoulli_.float
aten::normal_
aten::ones_like
aten::randn_like
aten::randint_like
aten::slice_scatter
aten::split_with_sizes
aten::split_with_sizes_copy
aten::uniform_
aten::zero_
```
- Missing tensor meta data:
```
aten::detach
aten::_to_copy
aten::bucketize.Tensor
aten::stack
aten::clone
aten::cat
aten::split.Tensor
aten::slice.Tensor
aten::copy_
aten::select.int
aten::slice_scatter
aten::split_with_sizes
aten::split_with_sizes_copy
aten::zero_
```

- Missing input_specs
```
aten::detach
aten::bernoulli
aten::_to_copy
aten::clone
aten::equal
aten::slice.Tensor
aten::copy_
aten::bernoulli_.float
aten::normal_
aten::slice_scatter
aten::uniform_
aten::zero_
```
Not that many ops' cost are missing since many ops are sharing the same strategy. I think it's doable that we can add those fields into the code directly. Though I hope we can remove the `redistribute_cost` hard code from all strategy registration to make the code cleaner. Any suggestions?

zpcore committed:
Support choice of register through wrapper

wanchaol commented:
@zpcore I think let's just fill the redistribute cost for each of them that are missing. One thing that's beneficial for just fixing those who are missing is that the redistribute cost would be generated together with sharding prop, so it eliminates another round of CPU overhead (although it might be small, but DTensor is pretty sensitive to CPU overhead).

zpcore commented:
I think auto cost generation is still meaningful that we can move it into the unit test to verify the redistribution cost computation correctness. Currently I have noticed several places that we incorrectly computed the cost e.g., 
https://github.com/pytorch/pytorch/blob/7cfd054075b0713642fab8dc4313fc1e5d992599/torch/distributed/tensor/_ops/_tensor_ops.py#L399-L400
https://github.com/pytorch/pytorch/blob/7cfd054075b0713642fab8dc4313fc1e5d992599/torch/distributed/tensor/_ops/_tensor_ops.py#L194

I prefer to have the unit test to verify the redistribute cost from strategy registration. What do you guys think?

wconstab commented:
Personally I love the idea of deleting all the code relating to manual creation of cost lists.  

I also think the cost lists are structured in a way that is not too easy to interpret.  (Each OpSpec has costs for inputs that refer to other OpSpecs that are all part of some parent Strategy).

I think we could address @wanchaol's comment about overheads.

> One thing that's beneficial for just fixing those who are missing is that the redistribute cost would be generated together with sharding prop, so it eliminates another round of CPU overhead (although it might be small, but DTensor is pretty sensitive to CPU overhead).

What if we made the cost generation happen in a way that is LRU cached and it should happen during sharding propagation anyway, so it would not happen _again_ at runtime
